### PR TITLE
Put everything on default_grid if not multigrid

### PR DESCRIPTION
**Pull request by coditva**
Tasks:
- [x] Fix `grid_move_line`. (in progress)
- [x] Find out if anymore functions need to be changed.

### CODE REVIEW

File: src/nvim/screen.c [hunk 1]
```diff
@@ -4344,6 +4337,18 @@ static void grid_move_line(ScreenGrid *grid, int row, int coloff, int endcol,
...
 
+  // If UI is not externalized, merge the contents of global and window grids
+  if (!ui_is_external(kUIMultigrid) && grid != &default_grid) {
+
+    //TODO(utkarshme): Change back to 1 (or 2)
+    my_grid_alloc(&new_grid, 3, endcol);
+
+    grid_buffer_cpy(&new_grid, &default_grid, row, endcol);
+    grid_buffer_cpy(&new_grid, grid, row, endcol);
+    grid = &new_grid;
```
Comment (coditva) [inline src/nvim/screen.c]: @bfredl I really need help on this. I can't seem to figure out what I keep doing wrong. Help! :sob:

Comment (bfredl) [reply src/nvim/screen.c]: Sorry, I can't quite see what `new_grid` is for. The "quickfix" solution I had in mind (for correctness only, not caring about allocation) was something like
```diff
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index 9b01d5c97..6b509b676 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -4299,6 +4299,14 @@ static void grid_move_line(ScreenGrid *grid, int row, int coloff, int endcol,
   if (!ui_is_external(kUIMultigrid) && grid != &default_grid) {
     row += grid->OffsetRow;
     coloff += grid->OffsetColumn;
+#define OFF_FROM(g) ((g).Rows*(g).Columns)
+    memcpy(default_grid.ScreenLines+OFF_FROM(default_grid),
+           grid->ScreenLines+OFF_FROM(*grid),
+           sizeof(schar_T)*grid->Columns);
+    memcpy(default_grid.ScreenAttrs+OFF_FROM(default_grid),
+           grid->ScreenAttrs+OFF_FROM(*grid),
+           sizeof(sattr_T)*grid->Columns);
+
     grid = &default_grid;
   }

```

In my testing this works with horizontal splits, but not vertical splits for some reason. (I would think the initial "from" offset is independent of x position, but I could be wrong).

Comment (coditva) [reply src/nvim/screen.c]: I was using `new_grid` onto which I copied data from both grids. That was redundant I see.
The patch makes things much better but segfaults even in horizontal splits if you scroll in the second split.
But this sure got me started and I'll see if I can work with it. Thank you!! :pray:

**Comment by bfredl** [comment]
After some minor corrections to offsets this seems to work pretty well:
```diff
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index 47a4fefad..3ea2cc198 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -5111,7 +5111,7 @@ win_redr_custom (
           *wp->w_p_stl == NUL ? 0 : OPT_LOCAL);
     }
 
-    col += wp->w_winrow;
+    col += wp->w_wincol;
   }
 
   if (maxwidth <= 0)
@@ -6327,6 +6327,7 @@ int grid_ins_lines(ScreenGrid *grid, int row, int line_count, int end,
   if (!ui_is_external(kUIMultigrid) && grid != &default_grid) {
     row += grid->OffsetRow;
+    end += grid->OffsetRow;
     col += grid->OffsetColumn;
     grid = &default_grid;
   }
 
@@ -6385,6 +6386,7 @@ int grid_del_lines(ScreenGrid *grid, int row, int line_count, int end,
   if (!ui_is_external(kUIMultigrid) && grid != &default_grid) {
     row += grid->OffsetRow;
+    end += grid->OffsetRow;
     col += grid->OffsetColumn;
     grid = &default_grid;
   }
```

Would there be any difficulties to merge this into the main PR? https://github.com/neovim/neovim/pull/8455

**Comment by coditva** [comment]
@bfredl I don't understand how I missed the `endcol` thing. I was sure I fixed it in some commit. :weary:
I don't thing there should be any more problems. I'll merge it and then rebase the tests branch and merge it when all tests pass (at least locally). :smile:

**Pull request merged**